otaserver/coap.py

```diff
@@ -64,24 +97,48 @@ def __init__(self, fw_path, port=COAP_PORT):
         self.port = port
         self.fw_path = fw_path
         root_coap = resource.Site()
-        root_coap.add_resource(('version', ), FirmwareVersionResource(self))
-        root_coap.add_resource(('firmware', ), FirmwareBinaryResource(self))
+        self.appids = {}
+        firmwares = os.listdir(self.fw_path)
```

Comment:
There's a bootstrap problem with this strategy: if the firmware path is empty, no resources CoAP is created and this is typically the case the very first time the application is started.
Since the CoAP controller is instanciated once, even if you upload a firmware after the application is started the CoAP resource will not exists and the node won't be able to ask for new version (error 4.04)
I think you need a PeriodicTask that checks if there are firmwares and add the resources in the `root_coap`.